File: 113/fp.cpp
#include <cmath>
#include <iostream>

using namespace std;

double calculateEuclideanDistance(double lat1, double lon1, double lat2, double lon2) {
   return round(sqrt(pow(lat1 - lat2, 2) + pow(lon1 - lon2, 2)), 2);
}

void merge(double arr[][5], int left[], int leftSize, int right[], int rightSize) {
   int i = 0, j = 0, k = 0;
   while (i < leftSize && j < rightSize) {
       if (left[i] < right[j]) {
           arr[k][4] = left[i];
           i++;
       } else {
           arr[k][4] = right[j];
           j++;
       }
       k++;
   }

   while (i < leftSize) {
       arr[k][4] = left[i];
       i++;
       k++;
   }

   while (j < rightSize) {
       arr[k][4] = right[j];
       j++;
       k++;
   }
}

void mergeSort(double arr[][5], int size) {
   if (size < 2) {
       return;
   }

   int mid = size / 2;
   int left[mid];
   int right[size - mid];

   for (int i = 0; i < mid; i++) {
       left[i] = arr[i][4];
   }

   for (int i = mid; i < size; i++) {
       right[i - mid] = arr[i][4];
   }

   mergeSort(arr, mid);
   mergeSort(arr, size - mid);

   merge(arr, left, mid, right, size - mid);
}

void calculateKNN(double arr[][5], int size, double lat, double lon) {
   for (int i = 0; i < size; i++) {
       arr[i][4] = calculateEuclideanDistance(arr[i][1], arr[i][2], lat, lon);
   }
   mergeSort(arr, size);
}

string hasYes(double arr[][5], int size, int n = 0) {
   if (n >= 2) {
       return "Yes";
   } else if (size == 0) {
       return "No";
   } else if (arr[0][3] == 1) {
       return hasYes(arr, size - 1, n + 1);
   } else {
       return hasYes(arr, size - 1, n);
   }
}

int main() {
   const int size = 15;
   double data[size][5] = {
       {1, 54, 97, 1},
       {2, 99, 91, 0},
       {3, 61, 53, 1},
       {4, 43, 95, 0},
       {5, 60, 10, 1},
       {6, 40, 74, 0},
       {7, 16, 92, 1},
       {8, 8, 8, 0},
       {9, 90, 37, 1},
       {10, 13, 32, 0},
       {11, 85, 75, 1},
       {12, 98, 23, 0},
       {13, 94, 74, 1},
       {14, 54, 38, 0},
       {15, 36, 42, 1}
   };

   double lat = 52, lon = 25;

   calculateKNN(data, size, lat, lon);
   string result = hasYes(data, 3);

   cout << "Result: " << result << endl;

   return 0;
}

File: 113/fp.py
import math
def merge(a, b):
    if len(a) == 0:
        return b
    elif len(b) == 0:
        return a
    elif float(a[0][4]) < float(b[0][4]):
        return [a[0]] + merge(a[1:], b)
    else:
        return [b[0]] + merge(a, b[1:])

def mergesort(x):
    if len(x) < 2:
        return x
    else:
        h = len(x) // 2
        return merge(mergesort(x[:h]), mergesort(x[h:]))

def kNN(data, lat, long):
    if len(data) == 0:
        return []
    else:
        k = round(math.sqrt(pow((data[0][1] - lat), 2) + pow((data[0][2] - long), 2)), 2)
        return [data[0] + [k]] + kNN(data[1:], lat, long)

def hasYes(data, n = 0):
    if len(data) == 0:
        if n<2:
            return "No"
        else:
            return "Yes"
    elif data[0][3] == "Yes":
        return hasYes(data[1:], n+1)
    else:
        return hasYes(data[1:], n)

def process(data, lat, long):
    knn = kNN(data, lat, long)
    sorted = mergesort(knn)
    result = hasYes(sorted[0:3])
    return sorted, result

data = [
    [1, 54, 97, "Yes"],
    [2, 99, 91, "No"],
    [3, 61, 53, "Yes"],
    [4, 43, 95, "No"],
    [5, 60, 10, "Yes"],
    [6, 40, 74, "No"],
    [7, 16, 92, "Yes"],
    [8, 8, 8, "No"],
    [9, 90, 37, "Yes"],
    [10, 13, 32, "No"],
    [11, 85, 75, "Yes"],
    [12, 98, 23, "No"],
    [13, 94, 74, "Yes"],
    [14, 54, 38, "No"],
    [15, 36, 42, "Yes"]
]
lat, long = 52, 25

sorted_data, result = process(data, lat, long)
print(data[0:3])
print("(Yes / No)  ? :", result)